README.md
# spliton

Split audio or video files into chunks.

# Usage (Examples)

```js
const outDir = '/home/user/cut-videos';

const splitEntries = [
  {
    startTime: '00:05',
    endTime: '00:10',
  },
  {
    startTime: '00:10',
    endTime: '00:30',
  },
];

const sourceFile = '/home/user/videos/sample.mp4';

const data = await new Spliton()
  .setOutDir(outDir)
  .add(splitEntries)
  .run(sourceFile);

// Data will be returned in such format:
/**
  [
      {
          "filename": "4e1ac1bc-e87c-4347-ba94-f44be8ee7c92.mp4",
          "ss": "00:05",
          "t": "00:10"
      },
      {
          "filename": "e6706df5-e29c-4974-b902-0d4052d5141e.mp4",
          "ss": "00:10",
          "t": "00:30"
      }
  ]
 */
```

# Dependencies

- Requires ffmpeg to be available.
- To set custom ffmpeg path:

```js
const customFFmpegPath = '/usr/bin/ffmpeg';
const sourceFile = '/home/user/videos/sample.mp4';

const data = await new Spliton()
  .setCustomFFmpegPath(customFFmpegPath);
  .setOutDir(outDir)
  .add(splitEntries)
  .run(sourceFile);
```

- Or set ffmpeg binary to be available under **PATH** environment variable.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.6](https://github.com/social-media-tools/spliton/compare/v0.0.5...v0.0.6) (2023-04-08)


### Bug Fixes

* added support for audio file splitting ([29dd4fa](https://github.com/social-media-tools/spliton/commit/29dd4fac60bb0349b4ab8993d8a761f3fa32e282))

### [0.0.5](https://github.com/social-media-tools/spliton/compare/v0.0.4...v0.0.5) (2023-02-19)


### Bug Fixes

* updated package information ([3d3a792](https://github.com/social-media-tools/spliton/commit/3d3a792f4b000346c455b09a7913ef6219fe5073))

### [0.0.4](https://github.com/social-media-tools/spliton/compare/v0.0.3...v0.0.4) (2023-02-18)


### Bug Fixes

* updated unit test ([34d3d44](https://github.com/social-media-tools/spliton/commit/34d3d442bdc360e774e85481e081d6f1f4a414b5))
* updated unit tests ([116b662](https://github.com/social-media-tools/spliton/commit/116b662e99ee7e972f0f57d7833aa6757d1e8e0a))
* use shell from ENV ([a7c3862](https://github.com/social-media-tools/spliton/commit/a7c3862191c560d2797597db4fc964d7134b43da))

### [0.0.2](https://github.com/social-media-tools/spliton/compare/v0.0.1...v0.0.2) (2023-02-16)


### Bug Fixes

* added command to build tar package ([608dc02](https://github.com/social-media-tools/spliton/commit/608dc02a1527b2ff0246e19228677067fa1d17ed))
* updated repository link ([75091b1](https://github.com/social-media-tools/spliton/commit/75091b16ed9f976e6dcdfc24d825a26e47f6912f))
* updated unit tests ([02c606f](https://github.com/social-media-tools/spliton/commit/02c606f5d554a8742e9eb03b0f08bd1f77f575d7))


src/lib/tests/spliton.test.ts
import fs from 'fs';
import path from 'path';

import test from 'ava';

import { Spliton } from '../spliton/spliton';
import { getAssetsDir } from '../utils';

test('Creates new video', async (t) => {
  const assetDir = getAssetsDir();
  const videoPath = path.join(assetDir, 'sample.mp4');
  await new Spliton()
    .add({
      startTime: '00',
      endTime: '02',
    })
    .add([
      {
        startTime: '00',
        endTime: '02',
      },
    ])
    .setOutDir(path.join(__dirname, 'temp'))
    .run(videoPath);
  t.pass();
});

test('Format time works as expected', (t) => {
  const time = Spliton.formatTime('00', '00');
  t.is(time, '00:00:00');
});

test('Allows to set ffmpeg binary path', (t) => {
  const spliton = new Spliton();
  spliton.setCustomFFmpegPath('/usr/bin/ffmpeg');
  t.is(spliton.ffmpegPath, '/usr/bin/ffmpeg');
});

test('Validates ffmpeg entries', async (t) => {
  const assetDir = getAssetsDir();
  const videoPath = path.join(assetDir, 'sample.mp4');
  try {
    await new Spliton()
      .add({
        startTime: '02',
        endTime: '01',
      })
      .run(videoPath);
  } catch (error: unknown) {
    const e = error as Error;
    t.is(e.message, 'invalid entry splits');
  }
});

test('Fails on invalid ffmpeg command', async (t) => {
  const assetDir = getAssetsDir();
  const videoPath = path.join(assetDir, 'sample.mp5');

  const data = await new Spliton()
    .add({
      startTime: '02',
      endTime: '04',
    })
    .setOutDir(path.join(__dirname, 'temp'))
    .run(videoPath);

  t.not(data[0].errorMessage, null);
});

test('Creates correct audio file when _isAudio is set to true and audio file is provided', async (t) => {
  const assetDir = getAssetsDir();
  const audioPath = path.join(assetDir, 'sample.mp3');
  const outputDir = path.join(__dirname, 'temp');

  fs.mkdirSync(outputDir, { recursive: true });

  const data = await new Spliton()
    .add({
      startTime: '00',
      endTime: '05',
    })
    .setOutDir(outputDir)
    .run(audioPath);

  t.is(data[0].errorMessage, undefined);
});


src/lib/spliton/spliton.ts
import { spawn } from 'child_process';
import crypto from 'crypto';
import path from 'path';

type SplitEntry = {
  startTime: string;
  endTime: string;
};

type FFMpegEntry = {
  filename: string;
  ss: string;
  t: string;
  errorMessage?: string;
  cmd?: string;
};

type ClockTime =
  | '00'
  | '01'
  | '02'
  | '03'
  | '04'
  | '05'
  | '06'
  | '07'
  | '08'
  | '09'
  | '10'
  | '11'
  | '12'
  | '13'
  | '14'
  | '15'
  | '16'
  | '17'
  | '18'
  | '19'
  | '20'
  | '21'
  | '22'
  | '23'
  | '24'
  | '25'
  | '26'
  | '27'
  | '28'
  | '29'
  | '30'
  | '31'
  | '32'
  | '33'
  | '34'
  | '35'
  | '36'
  | '37'
  | '38'
  | '39'
  | '40'
  | '41'
  | '42'
  | '43'
  | '44'
  | '45'
  | '46'
  | '47'
  | '48'
  | '49'
  | '50'
  | '51'
  | '52'
  | '53'
  | '54'
  | '55'
  | '56'
  | '57'
  | '58'
  | '59'
  | '60';
/**
 * Spliton API that allows to split video / audio input
 */
export class Spliton {
  ffmpegPath?: string;
  entries: SplitEntry[] = [];
  inputPath!: string;
  ext!: string;
  outDir!: string;

  private get ffmpeg() {
    return this.ffmpegPath || 'ffmpeg';
  }

  private get _ext() {
    if (!this.ext) {
      this.ext = path.extname(this.inputPath);
    }
    return this.ext;
  }

  private get randomUUID() {
    return crypto.randomUUID();
  }

  private getFfmpegEntries() {
    const ffmpegEntries: FFMpegEntry[] = this.entries.map((entry) => ({
      filename: `${this.randomUUID}${this._ext}`,
      ss: entry.startTime,
      t: entry.endTime,
    }));
    return ffmpegEntries;
  }

  private getCommand(entry: FFMpegEntry) {
    const input = `-i ${this.inputPath}`;
    let command = `${this.ffmpeg} `;
    const ss = `-ss ${entry.ss}`;
    const t = `-t ${entry.t}`;

    const map = `-map 0`;

    // Check if the input file is an audio file

    const outFile = path.join(this.outDir, entry.filename);
    const isAudio = ['.mp3', '.wav', '.m4a'].includes(this._ext);

    if (isAudio) {
      command += ` ${ss} ${t} ${input} ${map} -c:a copy ${outFile}`;
    } else {
      const c = `-c copy`;
      command += ` ${ss} ${t} ${input} ${c} ${map} ${outFile}`;
    }

    return command;
  }

  private validateEntries() {
    const entryErrors = [];
    for (const entry of this.entries) {
      let reason = '';
      if (entry.startTime > entry.endTime) {
        reason = 'Start time cannot be higher than end time';
      }
      // if (entry.startTime < 0 || entry.endTime < 0) {
      //   reason = 'Must be timestamp';
      // }
      if (reason) {
        entryErrors.push({
          entry,
          reason,
        });
      }
    }
    if (entryErrors.length !== 0) {
      throw new Error('invalid entry splits');
    }
  }

  setCustomFFmpegPath(ffmpegPath: string) {
    this.ffmpegPath = ffmpegPath;
    return this;
  }

  setOutDir(outDir: string) {
    this.outDir = outDir;
    return this;
  }

  add(entries: SplitEntry | SplitEntry[]) {
    const data = Array.isArray(entries) ? entries : [entries];
    this.entries = [...this.entries, ...data];
    return this;
  }

  async run(inputPath: string) {
    this.inputPath = inputPath;
    this.validateEntries();
    const ffmpegEntries = this.getFfmpegEntries();
    for await (const entry of ffmpegEntries) {
      entry.cmd = this.getCommand(entry);
      try {
        await this.callFfmpeg(entry.cmd);
      } catch (error) {
        const e = error as Error;
        entry.errorMessage = e.message;
      }
    }
    return ffmpegEntries;
  }

  private callFfmpeg(ffmpegCmd: string) {
    return new Promise((resolve, reject) => {
      const proc = spawn(ffmpegCmd, { shell: process.env.SHELL });

      let err = '';
      proc.stderr.on('data', (data) => (err += data.toString()));

      /* istanbul ignore next */
      proc.once('error', (err) => {
        return reject(err);
      });
      /* istanbul ignore next */
      proc.once('close', (errorCode) => {
        if (errorCode === 0) {
          return resolve(true);
        }
        reject(new Error(err));
      });
    });
  }

  static formatTime(mm: ClockTime, ss: ClockTime) {
    return `00:${mm}:${ss}`;
  }
}
